lib/segment/src/segment_constructor/segment_builder.rs

```diff
@@ -112,51 +125,133 @@ impl SegmentBuilder {
+                            entry.segment_index = segment_index;
+                            entry.version = version;
+                        }
+                    })
+                    .or_insert_with(|| {
+                        let internal_id = segment.get_internal_id(external_id).unwrap();
+                        PositionedPointMetadata {
+                            segment_index,
+                            internal_id,
+                            version,
+                        }
+                    });
+            }
+        }
 
-        let other_id_tracker = other.id_tracker.borrow();
-        let other_vector_storages: HashMap<_, _> = other
-            .vector_data
+        let payloads: Vec<_> = segment_guards
             .iter()
-            .map(|(vector_name, vector_data)| {
-                (vector_name.to_owned(), vector_data.vector_storage.borrow())
+            .map(|i| i.payload_index.borrow())
+            .collect();
+
+        let mut points_to_insert: Vec<_> = merged_points
+            .into_iter()
+            .map(|(external_id, pd)| {
+                let mut hasher = AHasher::default();
+
+                // If payload key is set, sort the points to add. For this we fetch the payload and
+                // hash it.
+                let payload_index = self.defragment_key.as_ref().and_then(|defragment_key| {
+                    payloads[pd.segment_index].field_indexes.get(defragment_key)
+                });
+                if let Some(payload_indexes) = payload_index {
+                    // Try to find an index we can use for hashing.
+                    for payload_index in payload_indexes {
+                        match payload_index {
+                            FieldIndex::IntMapIndex(index) => {
+                                if let Some(numbers) = index.get_values(pd.internal_id) {
+                                    for number in numbers {
+                                        number.hash(&mut hasher);
+                                    }
+                                }
+                                break;
+                            }
+                            FieldIndex::KeywordIndex(index) => {
+                                if let Some(keywords) = index.get_values(pd.internal_id) {
+                                    for keyword in keywords {
+                                        keyword.hash(&mut hasher);
+                                    }
+                                }
+                                break;
+                            }
+                            _ => (),
+                        }
+                    }
+                }
+
+                PositionedPointMetadataOrderable {
+                    segment_index: pd.segment_index,
+                    internal_id: pd.internal_id,
+                    external_id,
+                    version: pd.version,
+                    hash: hasher.finish(),
```

Comment:
We need to convert the HashMap to a Vec to be able to sort it, so I added hashing into this process.
This way we don't need to allocate another Vec again for the keys and indices in [sort_by_cached_key](https://doc.rust-lang.org/src/alloc/slice.rs.html#360).
However this is still a tradeoff since we now store the hash values longer.
We probably need to profile it to figure out which way actually uses less memory and it might also depend on the amount of entries too but what do you think?